### PR TITLE
[GraphQL] Add Dataloaders

### LINKED ISSUE #3075: [GraphQL] Implement Facebook Dataloader
https://graphql-dotnet.github.io/docs/guides/dataloader/

cc/ @RajasekarPR

### PR DESCRIPTION
**Pull request by carlwoodhouse**
fixes #3075 

This add's the dataloader extension points for graphql; which helps to eliminate the 1 * n query issue in graphql; when resolving fields that need to make a db call (basically for each record bought back, it will need to another query to resolve that field.

More info at https://graphql-dotnet.github.io/docs/guides/dataloader

As part of this PR i've had to switch back to `ParallelExecutionStrategy`, we originally opted out of this because `ISession` is not threadsafe (https://github.com/OrchardCMS/OrchardCore/issues/3029); but dataloaders just wont work without it. (see: https://github.com/graphql-dotnet/graphql-dotnet/issues/1191, etc for details) In order to combat the original issue i've made a new async resolver called `LockedAsyncFieldResolver` which should be used when needing to access ISession within a field resolver when *not* using dataloaders; this will basically lock the resolver to a semaphore per request (not per thread). This wont hinder perf anymore then using `SerialExecutionStrategy` like we did previously, infact it should increase it .. as any fields that dont need ISession (most of them) will resolve in Parallel :)

* Switched back to ParallelExecutionStrategy
* Added a new threadsafe resolver for fields that need to access ISession and dont use dataloaders
* Added the dataloader execution listener
* Added an extention for adding a preconfigured loader + query for published contentitems by id(s)
* Wired up the contentpickerfield with said dataloader
* added data loader to for listpart

*Breaking Change*
People who have implemented customgraphtypes that need to access to ISession in their resolvers will need to switch to the `LockedAsyncFieldResolver`.

### CODE REVIEW

File: src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Extensions/DataLoaderExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+            var accessor = serviceProvider.GetRequiredService<IDataLoaderContextAccessor>();
+            var session = serviceProvider.GetService<ISession>();
+
+            return accessor.Context.GetOrAddBatchLoader<string, ContentItem>("GetPublishedContentItemsById", ci => LoadPublishedContentItems(ci, session));
+        }
+
+        private static async Task<IDictionary<string, ContentItem>> LoadPublishedContentItems(IEnumerable<string> contentItemIds, ISession session)
+        {
+            var contentItemsLoaded = await session.Query<ContentItem, ContentItemIndex>(y => y.ContentItemId.IsIn(contentItemIds) && y.Published).ListAsync();
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Extensions/DataLoaderExtensions.cs]: It needs to be ordered with the same order as the key, you lost this piece:
```
contentItems.OrderBy(c => Array.IndexOf(x.Source.ContentItemIds, c.ContentItemId));
```

File: src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Extensions/DataLoaderExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+            return contentItemsLoaded.ToDictionary(k => k.ContentItemId, v => v);
+        }
+    }
+}
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Extensions/DataLoaderExtensions.cs]: Newline

File: src/OrchardCore/OrchardCore.Apis.GraphQL.Abstractions/Extensions/DataLoaderExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+            return Task.WhenAll(tasks);
+        }
+
+        public static Task<T[]> LoadAsync<TKey, T>(this IDataLoader<TKey, T> dataLoader, params TKey[] keys)
+        {
+            return dataLoader.LoadAsync(keys.AsEnumerable());
+        }
+    }
+}
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Apis.GraphQL.Abstractions/Extensions/DataLoaderExtensions.cs]: newline

I will start commenting on these until most PRs are correct, sorry for the noise

File: src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Extensions/DataLoaderExtensions.cs (hunk 1 above)
Comment (carlwoodhouse) [reply src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Extensions/DataLoaderExtensions.cs]: The dataloader does this by default when it passes the keys to loadasync (maintains the order of the passed keys)

File: src/OrchardCore/OrchardCore.Apis.GraphQL.Abstractions/Extensions/DataLoaderExtensions.cs (hunk 3 above)
Comment (carlwoodhouse) [reply src/OrchardCore/OrchardCore.Apis.GraphQL.Abstractions/Extensions/DataLoaderExtensions.cs]: new line at the EOF yeah ?

File: src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Extensions/DataLoaderExtensions.cs (hunk 1 above)
Comment (sebastienros) [reply src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Extensions/DataLoaderExtensions.cs]: awesome, thanks

**Comment by sebastienros** [comment]
If you don't intend to solve the "todos" in the PR, move them to the issue

**Comment by carlwoodhouse** [comment]
@sebastienros i will fix them in the pr; but tomorrow as its late here now :) marked as draft just to make sure i was on the right track.

File: src/OrchardCore.Modules/OrchardCore.Lists/GraphQL/ListQueryObjectType.cs [hunk 4]
```diff
@@ -22,36 +26,36 @@ public ListQueryObjectType(IStringLocalizer<ListQueryObjectType> T)
...
+                                     .With<ContentItemIndex>(ci => ci.Published)
+                                     .With<ContainedPartIndex>(cp => cp.ListContentItemId.IsIn(contentItemIds))
+                                     .OrderBy(o => o.Order)
+                                     .ListAsync();
+
+            return query.ToLookup(k => k.As<ContainedPart>().ListContentItemId);
+        }
     }
-}
+}
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Lists/GraphQL/ListQueryObjectType.cs]: nit

Comment (carlwoodhouse) [reply src/OrchardCore.Modules/OrchardCore.Lists/GraphQL/ListQueryObjectType.cs]: dammit! dont mind doing it but can i ask the reasoning ;p

**Comment by carlwoodhouse** [comment]
ok this is gonna be more work then anticipated, the dataloaders dont work right when using serial execution strategy which we need to stop session being shared across resolvers in parallel (https://github.com/OrchardCMS/OrchardCore/issues/3029)

**Comment by carlwoodhouse** [comment]
@sebastienros pushed changes around what we discussed (switching back to ParallelExecutionStrategy) i've achieved the session locking using a new threadsade field resolver for those fields that need to access ISession, tried a lot of different scenarios and i can no longer break it and dataloading works perfect.

The only thing is it might be a potential breaking change for people who have written their own custom graphtypes / schema builders that they'd need to switch to dataloaders or the new resolver.

Heres some pretty pictures ;) 

For this query ....
image

Sql profiler before ...
image

Sql profiler after
image

You can see the two root queries remain (they don't and probably shouldn't use dataloaders) but all the n+1 queries on the list have condensed into one query :) even across both queries. Even on my pretty small dataset here, you can see the potential perf gain is massive.

**Comment by sebastienros** [comment]
I wanted to be clever and only lock on the resources a resolver would need, but in the end
- they mostly all rely on session
- there could be some calls to services that hide what could be locked (liquid templates for instance)

So I think your solution is best. I will add some comments in the code though for later.

**Pull request merged**